plugins/amdgpu/amdgpu_plugin_topology.c

```diff
@@ -0,0 +1,1104 @@
+		return true;
+	}
+	return false;
+
+}
+
+bool iolink_match(struct tp_iolink *src, struct tp_iolink *dest)
+{
+	if (!src->valid)
+		return true;
+
+	if (NODE_IS_GPU(src->node_to) != NODE_IS_GPU(dest->node_to))
+		return false;
+
+	/* XGMI link can replace PCIE links */
+	if (src->type == TOPO_IOLINK_TYPE_XGMI && dest->type == TOPO_IOLINK_TYPE_PCIE)
+		return false;
+
+	/* bi-directional links can replace uni-directional links */
+	if (src->peer != NULL && dest->peer == NULL)
+		return false;
+
+	return true;
+}
+
+bool map_device(struct tp_system *src_sys, struct tp_system *dest_sys, struct tp_node *src_node,
+		struct tp_node *dest_node, struct device_maps *maps, struct device_maps *new_maps)
+{
+	struct tp_iolink *src_iolink;
+
+	pr_debug("Evaluating mapping nodes [0x%04X -> 0x%04X]\n", src_node->gpu_id, dest_node->gpu_id);
+
+	if (!device_properties_match(src_node, dest_node)) {
+		pr_debug("[0x%04X -> 0x%04X] Device properties do not match\n", src_node->gpu_id, dest_node->gpu_id);
+		return false;
+	}
+
+	if (src_node->num_valid_iolinks > dest_node->num_valid_iolinks) {
+		pr_debug("[0x%04X -> 0x%04X] Mismatch between number of iolinks\n",
+				src_node->gpu_id, dest_node->gpu_id);
+		return false;
+	}
+
+	list_for_each_entry(src_iolink, &src_node->iolinks, listm) {
+		/* Go through list of iolinks to CPU and compare them */
+
+		if (!NODE_IS_GPU(src_iolink->node_to)) {
+			bool matched_iolink = false;
+			/* This is a iolink to CPU */
+			pr_debug("Found link to CPU node:%02d\n", src_iolink->node_to->id);
+
+			uint32_t dest_cpu_node_id;
+			dest_cpu_node_id = maps_get_dest_cpu(maps, src_iolink->node_to->id);
+			if (dest_cpu_node_id == INVALID_CPU_ID)
+				dest_cpu_node_id = maps_get_dest_cpu(new_maps, src_iolink->node_to->id);
+
+			if (dest_cpu_node_id == INVALID_CPU_ID)  {
+				struct tp_iolink *dest_iolink;
+				list_for_each_entry(dest_iolink, &dest_node->iolinks, listm) {
+					if (iolink_match(src_iolink, dest_iolink) &&
+						!maps_dest_cpu_mapped(maps, dest_iolink->node_to->id) &&
+						!maps_dest_cpu_mapped(new_maps, dest_iolink->node_to->id)) {
+						if (!maps_add_cpu_entry(new_maps, src_iolink->node_to->id, dest_iolink->node_to->id))
+							/* This is a critical error because we are out of memory */
+							return false;
+
+						matched_iolink = true;
+						break;
+					}
+				}
+			} else {
+				pr_debug("Existing CPU mapping found [%02d-%02d]\n", src_iolink->node_to->id, dest_cpu_node_id);
+				/* Confirm that the link to this CPU is same or better */
+
+				struct tp_iolink *dest_iolink = node_get_iolink_to_node_id(dest_node, src_iolink->type, dest_cpu_node_id);
+				if (dest_iolink && iolink_match(src_iolink, dest_iolink))
+					matched_iolink = true;
+			}
+			if (!matched_iolink) {
+				pr_debug("[0x%04X -> 0x%04X] Mismatch between iolink to CPU\n", src_node->gpu_id, dest_node->gpu_id);
+				return false;
+			}
+		} else {
+			/* If GPUs have P2P-PCIe iolinks to this GPU, then at least one CPU will
```

Comment:
> You only parse io_links. Those do not include PCIe P2P links. They are only exposed under p2p_links on the DKMS branch. For now you can ignore these. But we are working on making that upstreamable so this will eventually land on amd-staging-drm-next. The existence of P2P PCIe links will depend on many factors. Just the existence of CPU->GPU links is not sufficient.
Yes, I put that comment because I felt that we were missing some cases here, but I did not have an easy way to test any code I would implement here. I added some words to the comment to make it more clear that there is probably more work to do here later.